tools/checklocks/analysis.go

```diff
@@ -548,23 +531,43 @@ func (pc *passContext) checkBasicBlock(fn *ssa.Function, block *ssa.BasicBlock,
 		}
 	}
 
-	// Analyze all successors.
 	for _, succ := range block.Succs {
-		// Collect possible return values, and make sure that the lock
-		// state aligns with any return value that we may have found
-		// above. Note that checkBasicBlock will recursively analyze
-		// the lock state to ensure that Releases and Acquires are
-		// respected.
-		if pls := pc.checkBasicBlock(fn, succ, lff, ls, seen); pls != nil {
-			if rls != nil && !rls.isCompatible(pls) {
+		entry := pre[succ]; 
+		pre[succ] = entry.join(pc, fn.Pos(), ls)
```

Comment:
I don't think it can be `rls`. `rls` is only non-nil iff `checkInstruction` encounters a `ssa.Return`. So if the last instruction was not a return, then rls is `nil` and does not represent the post lock state of the block. `checkInstruction` modifies `ls`, which is then used as the post lock state. I think returning `ls` and `rls` is clearer as well.

Reply:
Alternatively, we could return `ls` since that is the post state of the block regardless of whether it returns 🤔.